File: src/loom_web/schemas.py
"""Pydantic response schemas for the loom_web API."""

from __future__ import annotations

from pydantic import BaseModel


class ProjectSummary(BaseModel):
    """Lightweight project descriptor returned by GET /api/projects."""

    qid: str
    title: str
    repo: str | None
    default_branch: str | None
    archived: bool


class ItemRef(BaseModel):
    """Minimal cross-reference used in dependency/dependent lists."""

    qid: str
    type: str
    title: str
    status: str | None


class ItemNode(BaseModel):
    """One node in the project tree (no body).  Used by TreeResponse.items."""

    qid: str
    type: str
    title: str
    status: str | None
    assignee: str | None
    branch: str | None
    pr_url: str | None
    deps: list[str]
    children: list[str]


class TreeResponse(BaseModel):
    """Response shape for GET /api/projects/{project}/tree."""

    root: str
    items: list[ItemNode]


class ItemUpdate(BaseModel):
    """WebSocket payload for a changed item (record minus body).

    Broadcast to all ``/ws`` clients whenever a loom ``.md`` file is
    created or modified.  ``deleted`` is always ``False`` here; use
    :class:`ItemTombstone` for deletions.
    """

    qid: str
    type: str
    title: str
    status: str | None
    assignee: str | None
    branch: str | None
    pr_url: str | None
    tags: list[str]
    archived: bool
    deleted: bool = False


class ItemTombstone(BaseModel):
    """WebSocket payload for a deleted item.

    Broadcast when the item file has been removed and ``loom.get(qid)``
    raises :class:`loom.errors.NotFound`.
    """

    qid: str
    deleted: bool = True


class ItemDetail(BaseModel):
    """Full item descriptor returned by GET /api/items/{qid}."""

    qid: str
    type: str
    title: str
    status: str | None
    assignee: str | None
    branch: str | None
    pr_url: str | None
    tags: list[str]
    archived: bool
    body: str
    dependencies: list[ItemRef]
    dependents: list[ItemRef]
    children: list[str]

File: web/server/schemas.ts
/**
 * Response schemas for the loom-app Bun HTTP server.
 *
 * Mirrors src/loom_web/schemas.py — field names and nesting are identical
 * so the Bun server produces JSON shapes parity with the Python server.
 */

/** Sentinel version — lets tests import a runtime value to verify the module loads. */
export const SCHEMAS_VERSION = 1 as const;

// ---------------------------------------------------------------------------
// GET /api/projects → ProjectSummary[]
// ---------------------------------------------------------------------------

/** Lightweight project descriptor returned by GET /api/projects. */
export interface ProjectSummary {
  qid: string;
  title: string;
  repo: string | null;
  default_branch: string | null;
  archived: boolean;
}

// ---------------------------------------------------------------------------
// Shared cross-reference
// ---------------------------------------------------------------------------

/** Minimal cross-reference used in dependency/dependent lists. */
export interface ItemRef {
  qid: string;
  type: string;
  title: string;
  status: string | null;
}

// ---------------------------------------------------------------------------
// GET /api/projects/{project}/tree → TreeResponse
// ---------------------------------------------------------------------------

/** One node in the project tree (no body). Used by TreeResponse.items. */
export interface ItemNode {
  qid: string;
  type: string;
  title: string;
  status: string | null;
  assignee: string | null;
  branch: string | null;
  pr_url: string | null;
  deps: string[];
  children: string[];
  updated_at: string;
  created_at: string;
}

/** Response shape for GET /api/projects/{project}/tree. */
export interface TreeResponse {
  root: string;
  items: ItemNode[];
}

// ---------------------------------------------------------------------------
// GET /api/items/{qid} → ItemDetail
// ---------------------------------------------------------------------------

/** Full item descriptor returned by GET /api/items/{qid}. */
export interface ItemDetail {
  qid: string;
  type: string;
  title: string;
  status: string | null;
  assignee: string | null;
  branch: string | null;
  pr_url: string | null;
  tags: string[];
  archived: boolean;
  body: string;
  dependencies: ItemRef[];
  dependents: ItemRef[];
  children: string[];
}
